frame/democracy/src/lib.rs

```diff
@@ -642,6 +673,70 @@ decl_module! {
 			);
 			Self::deposit_event(RawEvent::Undelegated(who));
 		}
+
+		/// Veto and blacklist the proposal hash. Must be from Root origin.
+		#[weight = SimpleDispatchInfo::FixedNormal(10_000)]
+		fn clear_public_proposals(origin) {
+			ensure_root(origin)?;
+
+			<PublicProps<T>>::kill();
+		}
+
+		/// Register the preimage for an upcoming proposal. This doesn't require the proposal to be
+		/// in the dispatch queue but does require a deposit, returned once enacted.
+		#[weight = SimpleDispatchInfo::FixedNormal(100_000)]
+		fn note_preimage(origin, encoded_proposal: Vec<u8>) {
+			let who = ensure_signed(origin)?;
+			let proposal_hash = T::Hashing::hash(&encoded_proposal[..]);
+			ensure!(!<Preimages<T>>::exists(&proposal_hash), "preimage already noted");
+
+			let deposit = <BalanceOf<T>>::from(encoded_proposal.len() as u32)
+				.saturating_mul(T::PreimageByteDeposit::get());
+			T::Currency::reserve(&who, deposit)?;
+
+			let now = <system::Module<T>>::block_number();
+			<Preimages<T>>::insert(proposal_hash, (encoded_proposal, who.clone(), deposit, now));
+
+			Self::deposit_event(RawEvent::PreimageNoted(proposal_hash, who, deposit));
+		}
+
+		/// Register the preimage for an upcoming proposal. This requires the proposal to be
+		/// in the dispatch queue. No deposit is needed.
+		#[weight = SimpleDispatchInfo::FixedNormal(100_000)]
+		fn note_imminent_preimage(origin,
+			encoded_proposal: Vec<u8>,
+			when: T::BlockNumber,
+			which: u32
+		) {
+			let who = ensure_signed(origin)?;
+			let proposal_hash = T::Hashing::hash(&encoded_proposal[..]);
+			ensure!(!<Preimages<T>>::exists(&proposal_hash), "preimage already noted");
+			let queue = <DispatchQueue<T>>::get(when);
+			let item = queue.get(which as usize).and_then(|x| x.as_ref())
+				.ok_or("dispatch queue entry not found")?;
+			ensure!(item.0 == proposal_hash, "dispatch queue entry invalid");
+
+			let now = <system::Module<T>>::block_number();
+			<Preimages<T>>::insert(proposal_hash, (encoded_proposal, who.clone(), <BalanceOf<T>>::zero(), now));
+
+			Self::deposit_event(RawEvent::PreimageNoted(proposal_hash, who, Zero::zero()));
+		}
+
+		/// Remove an expired proposal preimage and collect the deposit.
+		#[weight = SimpleDispatchInfo::FixedNormal(10_000)]
+		fn reap_preimage(origin, proposal_hash: T::Hash) {
+			let who = ensure_signed(origin)?;
+
+			if let Some((_, old, deposit, then)) = <Preimages<T>>::get(&proposal_hash) {
+				let now = <system::Module<T>>::block_number();
+				if now >= then + T::EnactmentPeriod::get() + T::VotingPeriod::get() {
+					// allowed to claim the deposit.
+					let _ = T::Currency::repatriate_reserved(&old, &who, deposit);
```

Comment:
Brutal...

Reply:
So this is literal a race to get the deposit by everyone?

There seems to be no mechanism that a user who knows their proposal failed can try to claim their reserved balance back first...

Reply:
Maybe an authorization check, which skips the "date" check.

The owner of the data can call this function at any time, but if someone else wants to reap the rewards, they need to wait until after expiry.

Reply:
it's a race, but it's not brutal - the only reason you'd ever use have funds reserved is if you know for sure that the proposal will pass before it hits the enactment period; otherwise, you'll just use the `_imminent` variant.